### PR TITLE
jasmine: update spec and suite

### PR DESCRIPTION
**Pull request by UziTech**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/jasmine/jasmine/blob/main/src/core/Spec.js> <https://github.com/jasmine/jasmine/blob/main/src/core/Suite.js>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by UziTech** [comment]
TODO:
- [ ] Rebase on #51997 after it is merged
- [ ] Add tests

File: types/jasmine/index.d.ts [hunk 1]
```diff
@@ -905,36 +907,28 @@ declare namespace jasmine {
...
-    }
+        new (attrs: any): any;
 
-    interface XSpec {
-        id: number;
+        setSpecProperty: typeof setSpecProperty;
+        getResult(): SpecResult;
+        execute(onComplete: (error?: Error) => void, excluded?: boolean, failSpecWithNoExpectations?: boolean): void;
+        pend(message?: string): void;
+        status(excluded?: boolean, failSpecWithNoExpectations?: boolean): string;
```
Comment (UziTech) [inline types/jasmine/index.d.ts:920]: @sgravrock These spec functions are not documented but I think they could be helpful to certain applications using Jasmine. Just wondering which ones (if any) I should add to the Spec type?

File: types/jasmine/index.d.ts [hunk 2]
```diff
@@ -905,36 +907,28 @@ declare namespace jasmine {
...
-        suites(): Suite[];
-    }
-
-    interface XSuite {
-        execute(): void;
+        children: (Spec | Suite)[];
+        setSuiteProperty: typeof setSuiteProperty;
+        getResult(): SuiteResult;
+        pend(): void;
+        statue(): void;
```
Comment (UziTech) [inline types/jasmine/index.d.ts:931]: @sgravrock same here for undocumented suite functions.

File: types/jasmine/index.d.ts (hunk 1 above)
Comment (UziTech) [reply types/jasmine/index.d.ts:920]: at least `setSpecProperty` and `getResult` seem pretty useful.

Comment (sgravrock) [reply types/jasmine/index.d.ts:920]: Please don't add any of those, or the `Suite` members below. None of them are safe to call at arbitrary times. `execute` is particularly scary -- I can't think of any way for user code to call it without corrupting Jasmine's internal state. But even something like `setSpecProperty` is only safe to call on the currently executing spec. The corresponding public APIs have safeguards that these internal methods don't.

One of the big lessons of Jasmine 1.x is that almost everything is a breaking change when user code has direct access to `Spec` and `Suite`. Hiding those objects was one of the main goals of the 2.0 rewrite. They were later re-exposed via `topSuite`, which was probably a mistake, but we can at least avoid making the problem worse by keeping the documented API as narrow as possible.

Thanks for asking.

**Review by chivesrs — APPROVED** [review]

**Comment by UziTech** [comment]
Ready to merge

**Comment by StanislavKharchenko** [comment]
Hello!
Could you please share more info regarding removal of interfaces like CustomReporterResult, SuiteInfo, etc?
Is somewhere changelog or documenation on this?

**Comment by devoto13** [comment]
@StanislavKharchenko I believe they were renamed and slightly changed to more correctly describe the shape of objects passed to the Jasmine hooks (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52004/files#diff-344efb0ef17abfa10b5fd8c8152d7e2250795f7bf199d26d00bcd5e5f3f80495L894). If you use them as part of hooks it should be a matter of renaming the imports (or removing them altogether - as parameter types should be inferred automatically). Or do you have another use case for these interfaces?

**Comment by StanislavKharchenko** [comment]
@devoto13 
We're using our custom reporters which inherit base hooks "jasmineStarted", "specStarted" etc and have default parameters with specified arguments and its types. Like:

    public jasmineStarted(suiteInfo: jasmine.SuiteInfo): void {
        super.jasmineStarted(suiteInfo);
        //Our own iimplementation
    }

And we will have compilation error since interface "SuiteInfo" doesn't exist with type jasmine.
So, we need to make changes in accordance with last interface naming.

**Comment by devoto13** [comment]
I see.

@UziTech Maybe we should also export new interfaces under old names as deprecated to ease the migration? 
Properties seem to be backward compatible, so it should be pretty simple to do. E.g.

```
/** @deprecated use JasmineStartedInfo instead */
type SuiteInfo = JasmineStartedInfo;

/** @deprecated use SuiteResult or SpecResult instead */
type CustomReporterResult = SuiteResult & SpecResult;

/** @deprecated use JasmineStartedInfo instead */
type RunDetails = JasmineDoneInfo;
```

**Comment by coyoteecd** [comment]
@UziTech @sgravrock I ended up here after doing an update of @types/jasmine, which now gives me a bunch of compile errors because of the renamed interfaces (I am using jasmine together with jasmine-spec-reporter package, and that package is now broken because of the rename).

That deprecation notice would've been great.

**Pull request merged**